controlplane/kubeadm/controllers/helpers.go

```diff
@@ -246,3 +248,87 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	machinesToUpgrade := internal.NewFilterableMachineCollection()
+
+	now := metav1.Now()
+	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
+		oldMachines := c.Machines.Filter(machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter))
+		// Ignore deleted machines.
+		// TODO: This should be removed once MHC remediation is merged.
+		oldMachines = oldMachines.Filter(func(machine *clusterv1.Machine) bool {
+			return machine.DeletionTimestamp.IsZero()
+		})
```

Comment:
Why does this need to be removed, isn't it separate from remediation?

Reply:
It is but deletion timestamp will be checked before even coming to this method, at least it was like this in the open[ MHC PR](https://github.com/kubernetes-sigs/cluster-api/pull/3185)

Reply:
Let's see how the remediation PR shapes up